File: Tyche.Android/core/src/main/java/com/felipearpa/core/paging/CursorPage.kt
package com.felipearpa.core.paging

typealias MapFunc<TA, TB> = (TA) -> TB

data class CursorPage<TModel : Any>(
    val items: List<TModel>,
    val next: String?
) {

    fun <TOut : Any> map(mapFunc: MapFunc<TModel, TOut>): CursorPage<TOut> {
        return CursorPage(
            items = this.items.map(mapFunc),
            next = this.next
        )
    }
}

fun <T : Any> emptyCursorPage() = CursorPage(
    items = emptyList<T>(),
    next = null
)

File: Tyche.Backend/Felipearpa.Core/Felipearpa.Core/Paging/CursorPage.fs
namespace Felipearpa.Core.Paging

open System.Linq

type CursorPage<'T> = { Items: 'T seq; Next: string option }

module CursorPage =

    let empty<'T> () =
        { CursorPage.Items = Seq.empty<'T>
          Next = None }

    let map func page =
        { CursorPage.Items = page.Items |> Seq.map func
          Next = page.Next }

    let cast<'T> page =
        { CursorPage.Items = page.Items |> Seq.cast<'T>
          Next = page.Next }

    let fromSeq seq = { CursorPage.Items = seq; Next = None }

    let isEmpty<'T> (page: CursorPage<'T>) = page.Items.Count() = 0
